src/main/scala/is/hail/methods/Skat.scala

```diff
@@ -0,0 +1,318 @@
+        qtxwArray(i * k + j) = qtxwsi.data(j)
+        j += 1
+      }
+      i += 1
+    }
+
+    //compute the variance component score
+    var skatStat = 0.0
+    i = 0
+    while (i < m) {
+      skatStat += st(i).q
+      i += 1
+    }
+
+    val weightedGenotypes = new DenseMatrix[Double](n, m, xwArray)
+    val qtWeightedGenotypes = new DenseMatrix[Double](k, m, qtxwArray)
+    val GwGrammian = weightedGenotypes.t * weightedGenotypes
+    val QtGwGrammian = qtWeightedGenotypes.t * qtWeightedGenotypes
+
+    val SPG = new SkatModel(skatStat / (2 * sigmaSq))
+    SPG.computeLinearSkatStats(GwGrammian, QtGwGrammian)
+  }
+
+  def sparseResultOp(st: Array[SkatTuple[SparseVector[Double]]], sigmaSq: Double): SkatStat = {
+
+    val m = st.length
+    val n = st(0).xw.size
+    val k = st(0).qtxw.size
+
+    val xwArray = new Array[Double](m * n)
+    val qtxwArray = new Array[Double](m * k)
+
+    var j = 0;
+    var i = 0
+
+    while (i < m) {
+      val nnz = st(i).xw.used
+      val xwsi = st(i).xw
+      j = 0
+      while (j < nnz) {
+        val index = st(i).xw.index(j)
+        xwArray(i * n + index) = xwsi.data(j)
+        j += 1
+      }
+      i += 1
+    }
+
+    //add in non-zeros to QtGW array
+    i = 0
+    while (i < m) {
+      j = 0
+      val qtxwsi = st(i).qtxw
+      while (j < k) {
+        qtxwArray(i * k + j) = qtxwsi.data(j)
+        j += 1
+      }
+      i += 1
+    }
+
+    //compute the variance component score
+    var skatStat = 0.0
+    i = 0
+    while (i < m) {
+      skatStat += st(i).q
+      i += 1
+    }
+
+    val weightedGenotypes = new DenseMatrix[Double](n, m, xwArray)
+    val qtWeightedGenotypes = new DenseMatrix[Double](k, m, qtxwArray)
+    val GwGrammian = weightedGenotypes.t * weightedGenotypes
+    val QtGwGrammian = qtWeightedGenotypes.t * qtWeightedGenotypes
+
+    val SPG = new SkatModel(skatStat / (2 * sigmaSq))
+    SPG.computeLinearSkatStats(GwGrammian, QtGwGrammian)
+  }
+
+  def largeNResultOp[T <: Vector[Double]](st: Array[SkatTuple[T]], sigmaSq: Double): SkatStat = {
+    val m = st.length
+    val n = st(0).xw.size
+    val k = st(0).qtxw.size
+
+    val ZGrammianArray = new Array[Double](m * m)
+    val QtZGrammianArray = new Array[Double](m * m)
+
+    var i = 0
+    var j = 0
+
+    while (i < m) {
+      ZGrammianArray(i * m + i) = st(i).xw dot st(i).xw
+      j = 0
+      while (j < i) {
+        val ijdotprod = st(i).xw dot st(j).xw
+        ZGrammianArray(i * m + j) = ijdotprod
+        ZGrammianArray(j * m + i) = ijdotprod
+        j += 1
+      }
+      i += 1
+    }
+
+    i = 0
+    while (i < m) {
+      QtZGrammianArray(i * m + i) = st(i).qtxw dot st(i).qtxw
+      j = 0
+      while (j < i) {
+        val ijdotprod = st(i).qtxw dot st(j).qtxw
+        QtZGrammianArray(i * m + j) = ijdotprod
+        QtZGrammianArray(j * m + i) = ijdotprod
+        j += 1
+      }
+      i += 1
+    }
+
+    //compute the variance component score
+    var skatStat = 0.0
```

Comment:
This is the variance component score from the original paper. I could rename to include sum in the name, but I felt that would obfuscate the variable as the value is presented as an inner product in the paper.